src/Workspaces/Core/Portable/Utilities/EditDistance.cs

```diff
@@ -47,572 +68,581 @@ private static char[] ConvertToLowercaseArray(string text)
+            // without their final characters.  If their final characters are the same, they'll ahve the
+            // same edit distance.  If they are different, the edit distance will be greater.  Given 
+            // that we know the final edit distance is in the lower right, we can discover something 
+            // useful in the matrix.
+            //
+            //           s u n d a y
+            //      ----------------
+            //      |∞ ∞ ∞ ∞ ∞ ∞ ∞ ∞
+            //      |∞ 0 1 2 3 4 5 6
+            //    s |∞ 1 
+            //    a |∞ 2
+            //    t |∞ 3 `
+            //    u |∞ 4   `
+            //    r |∞ 5     `
+            //    d |∞ 6       `
+            //    a |∞ 7         `
+            //    y |∞ 8           *
+            //
+            // The slashes are the diagonal leading to the lower right.  The value in the lower right will
+            // be strictly equal to or greater than any value on this diagonal.  Thus, if that value 
+            // exceeds the threshold, we know we can stop immediately as the total edit distance must be
+            // greater than the threshold.
+            //
+            // We can use similar logic to avoid even having to examine more of the matrix when we
+            // have a threshold. First, consider the bottom diagonal.
+            // 
+            //           s u n d a y
+            //      ----------------
+            //      |∞ ∞ ∞ ∞ ∞ ∞ ∞ ∞
+            //      |∞ 0 1 2 3 4 5 6
+            //    s |∞ 1
+            //    a |∞ 2
+            //    t |∞ 3 `
+            //    u |∞ 4   `       x
+            //    r |∞ 5     `     |
+            //    d |∞ 6       `   |
+            //    a |∞ 7         ` |
+            //    y |∞ 8           *
+            //
+            // And then consider a point above that diagonal (indicated by x).  In the example
+            // above, the edit distance to * from + will be (x+4).  If, for example, threshold
+            // was '2', then it would be impossible for the path from 'x' to provide a good
+            // enough edit distance *ever*.   Similarly:
+            //
+            //           s u n d a y
+            //      ----------------
+            //      |∞ ∞ ∞ ∞ ∞ ∞ ∞ ∞
+            //      |∞ 0 1 2 3 4 5 6
+            //    s |∞ 1
+            //    a |∞ 2
+            //    t |∞ 3 `
+            //    u |∞ 4   `
+            //    r |∞ 5     `
+            //    d |∞ 6       `
+            //    a |∞ 7         `
+            //    y |∞ 8     y - - *
+            //
+            // Here we see that the final edit distance will be "y+3".  Again, if the edit 
+            // Distance threshold is less than 3, then could never contribute a path.
+            //
+            // So, if we had an edit distance threshold of 3, then the range around that
+            // bottom diagonal that we should consider checking is:
```

Comment:
I'll assume for the remainder of this that you have introduced the "top diagonal" idea